flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java

```diff
@@ -704,6 +716,66 @@ public void testResourceManagerConnectionAfterRegainingLeadership() throws Excep
 		}
 	}
 
+	private JobGraph createDataSourceJobGraph() {
+		final TextInputFormat inputFormat = new TextInputFormat(new Path("."));
+		final InputFormatVertex producer = new InputFormatVertex("Producer");
+		new TaskConfig(producer.getConfiguration()).setStubWrapper(new UserCodeObjectWrapper<InputFormat<?, ?>>(inputFormat));
+		producer.setInvokableClass(DataSourceTask.class);
+
+		final JobVertex consumer = new JobVertex("Consumer");
+		consumer.setInvokableClass(NoOpInvokable.class);
+		consumer.connectNewDataSetAsInput(producer, DistributionPattern.POINTWISE, ResultPartitionType.BLOCKING);
+
+		final JobGraph jobGraph = new JobGraph(producer, consumer);
+		jobGraph.setAllowQueuedScheduling(true);
+
+		return jobGraph;
+	}
+
+	/**
+	 * Tests the {@link JobMaster#requestNextInputSplit(JobVertexID, ExecutionAttemptID)}
+	 * validate that it will get same result for a different retry
+	 */
+	@Test
+	public void testRequestNextInputSplitWithDataSourceFailover() throws Exception {
+
+		final JobGraph dataSourceJobGraph = createDataSourceJobGraph();
+		testJobMasterAPIWithMockExecution(dataSourceJobGraph, (tdd, jobMaster) ->{
+			try{
+				final JobMasterGateway gateway = jobMaster.getSelfGateway(JobMasterGateway.class);
+
+				final TaskInformation taskInformation = tdd.getSerializedTaskInformation()
+					.deserializeValue(getClass().getClassLoader());
+				JobVertexID vertexID = taskInformation.getJobVertexId();
+
+				//get the previous split
+				SerializedInputSplit split1 = gateway.requestNextInputSplit(vertexID, tdd.getExecutionAttemptId()).get();
+
+				//start a new version of this execution
+				ExecutionGraph executionGraph = jobMaster.getExecutionGraph();
+				Execution execution = executionGraph.getRegisteredExecutions().get(tdd.getExecutionAttemptId());
+				ExecutionVertex executionVertex = execution.getVertex();
+
+				long version = execution.getGlobalModVersion();
+				gateway.updateTaskExecutionState(new TaskExecutionState(dataSourceJobGraph.getJobID(), tdd.getExecutionAttemptId(), ExecutionState.FINISHED)).get();
+				Execution newExecution = executionVertex.resetForNewExecution(System.currentTimeMillis(), version);
+
+				//get the new split
+				SerializedInputSplit split2 = gateway.requestNextInputSplit(vertexID, newExecution.getAttemptId()).get();
+
+				Assert.assertArrayEquals(split1.getInputSplitData(), split2.getInputSplitData());
+
+				//get the new split3
```

Comment:
Make sure you cover the case where input splits are exhausted.

Reply:
add more assert